### ISSUE TITLE
Add schema update + index reloading [JIRA: RIAK-1591]

### ISSUE THREAD

**Issue by rzezeski** [basho/yokozuna#130]
After a conversation with someone on IRC it came to my attention that Yokozuna should support schema updates + index reloading sooner rather than later.  The user was performing these basic steps that would better be handled by Yokozuna itself.
1. Modify the schema.
2. Copy the schema to /data/yz/<index name>/conf/<schema name>.xml on each node.
3. Reboot each Riak node to reload the schema.
4. New objects can use new fields.  New fields can be queried on.

This could be done in Yokozuan like so:
1. Modify schema.
2. Update in Yokozuna: `curl -X PUT -H 'content-type: application/json' 'http://host:port/yz/schema/<schema name>' --data-binary @<schema name>.xml`
3. Tell Yokozuna to reload the index: `curl -X POST -H 'content-type: text/plain' 'http://host:port/yz/index/<index name>' RELOAD`

This would use Solr's core RELOAD command and wouldn't require a restart.

**Comment by arobson** [comment]
@rzezeski - so I've been working on getting yokozuna support for an upcoming feature and having to bounce the cluster every time we add a schema or make a schema change is a real concern. Any chance this might land in the 2.0 release proper?

**Comment by rzezeski** [comment]
Yes, this should land in 2.0 proper.

**Comment by rzezeski** [comment]
Some thought needs to be giving to Solr's new managed schema API.

http://wiki.apache.org/solr/SchemaRESTAPI

Should Yokozuna try to mimic this? I'm not sure this plays well with Yokozuna's distributed bits.

Consideration for #219 must also be given. If schemaless mode is used the schema is updating automatically in Solr itself. How do Yokozuna and Solr keep a consistent view of what the schema is for an index?

**Comment by rzezeski** [comment]
The server now has the ability to reload indexes on the fly thanks to #260. But there is still no HTTP, PB, or riak-admin interface yet.

**Comment by krestenkrab** [comment]
This is a real issue for the usability of Yokozuna, as developing anything usually requires modifying the schema as you go along.  We really need at least HTTP access to this, so it can be scripted.

Also, it's a bit unclear how to push an updated schema.  If you just re-PUT an updated schema to the HTTP `/search/schema/schemaName` url then the schema does not get rewritten into the corresponding `indexName/conf` directory for indexes that use the schema.  That would seem like an obvious semantics, I was a bit baffled about this at least.

A related issue is that it would be nice to have a way to PUT a document, and get an error back if it doesn't conform to a schema in the back end.  Maybe not as a default behaviour -- as it would be really slow -- but as a way to avoid having to scavenge the `log/solr.log` files for errors when queries subsequently do not return the expected results.

In short, the edit-deploy-test cycle for working with a schema needs to be improved.  IMHO, this needs done before you can reasonably include Yokozuna in a release.  I would mark this issue as [must] for 1.0.0.

**Comment by rzezeski** [comment]
> This is a real issue for the usability of Yokozuna, as developing anything usually requires modifying the schema as you go along. We really need at least HTTP access to this, so it can be scripted.
> 
> Also, it's a bit unclear how to push an updated schema. If you just re-PUT an updated schema to the HTTP /search/schema/schemaName url then the schema does not get rewritten into the corresponding indexName/conf directory for indexes that use the schema. That would seem like an obvious semantics, I was a bit baffled about this at least.

The current admin API for Yokozuna is awkward, but it is
workable. Exposing the reload command would avoid the need create a
new index just to edit the schema but modifying the schema of an index
with existing data also has many issues. There is related discussion
in the following issues:
- #316
- #219
- #10

In short: mutating or removing fields will break your index. Adding
fields will not retroactively re-index all your previously written
objects. Modifying analyzer chains will break queries. This is not
solely a Yokozuna issue; the same problems exist in
[ElasticSearch](http://www.elasticsearch.org/blog/changing-mapping-with-zero-downtime/) and SolrCloud.

> A related issue is that it would be nice to have a way to PUT a document, and get an error back if it doesn't conform to a schema in the back end. Maybe not as a default behaviour -- as it would be really slow -- but as a way to avoid having to scavenge the log/solr.log files for errors when queries subsequently do not return the expected results.

I agree, this would be nice and would like to add something like
this. There is no time for 2.0.

> In short, the edit-deploy-test cycle for working with a schema needs to be improved. IMHO, this needs done before you can reasonably include Yokozuna in a release. I would mark this issue as [must] for 1.0.0.

I agree with everything but the last part. I have many ideas for
improvement but Riak 2.0 is in code freeze and there is not enough
time to rework the entire API. It's awkward to use right now but it is
possible. There are people who have been using Yokozuna with success
for months now. This is an area that needs much improvement but I
disagree that it should prevent it from being shipped in Riak 2.0. It
just may not be ready according to your standards; and I can respect
that. You are making reasonable requests but the time is simply not
there.

I hope to put out an RFC soon about my ideas for a better
administration API. It attempts to solve a lot of the issues around
mutating schemas, allow synchronous index creation, get rid of
separate notion of schema (an index has a schema, not a pointer to a
schema), allow testing of index/schema before committing, allow staged
planning of schema changes, and move everything into a riak-admin
sub-command to provide a unified interface. It's one of N things I'd
like to see improved in Yokozuna.

Thank you for taking the time to comment on this issue and bring up
your concerns.

**Comment by krestenkrab** [comment]
I can understand that. I'm sorry this was mostly negative feedback :-)  Once things are all aligned right, things worked just fine.  You just have to walk a narrow path as things are now, so we should make sure the "good path" is well documented.

**Comment by arobson** [comment]
@rzezeski  as of riak 2.0.0 beta 1, we're not seeing any evidence that this works. If we put a changed schema (with the same schema name), SOLR never seems to get the new file. The only steps I've managed to come up w/ are to go to the SOLR admin, unload the index core and then re-run our code that then successfully re-creates the index with the new schema file.

If that's where things are, that's fine, it's just not entirely clear what's supported from the sparse docs I'm seeing. If there is anyway to make changes to a schema and have the index use the new schema without blowing away the index, could you point me in that direction? Thanks!

**Comment by rzezeski** [comment]
@arobson How exactly did you try to reload the schema?  There is no HTTP/PB interface for this functionality and there won't be in 2.0.0 final.  The only way to reload the index on the fly is to run the following on every node.

``` erlang
rp(yz_index:reload(<<"index_name">>)).
```

If all is well then you should get `{ok, Nodes}` where `Nodes` is the list of nodes in your Riak cluster.  If something goes wrong you'll get `{error, Errors}` where `Errors` is a list of errors for each node that had an error.

**Comment by rzezeski** [comment]
@krestenkrab Sorry it took so long.  I finally put some of my ideas down for a better administration API.  I can't say any of the work will be done any time soon but it is at least a straw man that can be picked apart. See #403.

**Comment by Basho-JIRA** [comment]
Removing assignee (changed to "Unassigned"). 

**_[posted via JIRA by Derek Somogyi]**_

**Comment by pseudorandumb** [comment]
Any idea as to when this will make it in to a release? It's been sitting here for a while :D

**Comment by zeeshanlakhani** [comment]
@pseudorandumb we're definitely working on this, but had to place some other bugfixes and additions ahead at the moment. There's already been some work on the cli and yz to make this happen. 

It's definitely on my agenda. Thanks.

**Comment by pseudorandumb** [comment]
Awesome, thanks for the quick response @zeeshanlakhani.

**Comment by ghost** [comment]
I'm currently running into this issue. Discussed on the [mailing list](http://lists.basho.com/pipermail/riak-users_lists.basho.com/2015-March/016954.html). 

Correct me if I'm wrong, but in order to fix this manually, we must follow the steps outlined in the first comment & then basically re-store every affected key/value pair?

**Comment by Basho-JIRA** [comment]
PROD: Attempt to fix for RIAK 3.0 

**_[posted via JIRA by Seema Jethani]**_

**Comment by dmitrizagidulin** [comment]
@mtbrooks33 It sounds like your question was answered on the mailing list? Otherwise, let us know.

**Comment by ghost** [comment]
Correct. Thank you! I'm looking forward to this feature

**Comment by Basho-JIRA** [comment]
PROD: Attempt to fix for RIAK 3.0. Customer Shopkeep 

***[posted via JIRA by Seema Jethani]***

**Comment by Basho-JIRA** [comment]
Client also looking for this feature - https://basho.zendesk.com/agent/tickets/10898 

**_[posted via JIRA by Bryan Hunt]**_

**Comment by Basho-JIRA** [comment]
[~bhunt] yep... no immediate timeline, but we're getting close to getting more time to tackle this. 

**_[posted via JIRA by Zeeshan Lakhani]**_

**Comment by sdebnath** [comment]
Adding a reference to my gist in case someone is looking for how to reload indexes, esp when CRDTs are involved: https://gist.github.com/sdebnath/36c235e042cb35db7d1f

**Comment by Basho-JIRA** [comment]
Review for 2.2 

***[posted via JIRA by Seema Jethani]***

**Comment by Basho-JIRA** [comment]
Duplicate of https://bashoeng.atlassian.net/browse/RIAK-1591... leaving that one open. 

**_[posted via JIRA by Zeeshan Lakhani]**_

**Comment by zeeshanlakhani** [comment]
Reference: https://github.com/basho/yokozuna/issues/403

**Comment by TSiggs** [comment]
Re-indexing via wiping yz hash trees breaks when using custom extractors.  Not sure what the implications are for this issue, but may be related.  Replication details here: https://basho.zendesk.com/agent/tickets/11623

**Comment by zeeshanlakhani** [comment]
Thanks @TSiggs. Can you move this into a separate issue? This actually doesn't fit within this one which is for a more distinct/general reindexing plan. Please do and then delete this post from this thread if you can. Thanks.

**Comment by karlra** [comment]
I would also like to say that this would be a huge help to me as well, especially being able to change the schema on an existing index. I understand that triggering an index rebuild would be very expensive but being able to get it to at least start indexing new data with the updated schema would be a big help during development.

**Comment by AliakseiMat** [comment]
So, Is this feature ( update schema ) eventually available? And in what riak version?

**Comment by Guibod** [comment]
Indeed, that's needed.
How do you guys handle this ? You create a new schema, then a new index ?

**Comment by evrenesat** [comment]
We are handling schema updates with following module, not perfect but works for us: https://github.com/zetaops/pyoko/blob/develop/pyoko/db/schema_update.py

**Comment by lastzero** [comment]
Updating a search schema is such a basic, fundamental use case that nobody in his right mind would expect this to be missing from a stable database release. How do you test a schema, if you can not update it easily? Do you need to get it right the first time? Hope I got something wrong here...

**Comment by runesl** [comment]
@evrenesat Good work on the python schema updater!
I see you're even streaming through the old keys one at a time to reindex them after schema update. 
Another strategy I've been considering is to dirty the AAE-trees to force AAE to reindex all data after schema update.

It seems your approach of using the temp index means searching is unavailable during the schema update+data re-indexing process? Ideally, we could still use the old index while building the new index, but Riak doesn't seem to support indexing with multiple schemas+indices on a bucket.

**Comment by arvind-kalyan** [comment]
Hi, 

We had a use case where we needed to add a new field for indexing into an already created schema. We tried changing the schema file by doing `curl -XPUT $RIAK_HOST/search/schema/temp_core  -H'content-type:application/xml' --data-binary @temp_core_schema.xml`, but it wasn't reflecting on the admin console. Then tried updating directly in `/var/lib/riak/yz/temp_core/conf/temp_core_schema.xml`, it was getting reflected asap. Also, when we tried adding a value for that field in Riak object, it was getting indexed properly. 

We had added a "default" attribute for this new field, but the old documents were unmodified. Ideally it should have the default value against this field. 

Is this a good workaround? Please tell me if there are any pros/cons to this. Also would love to hear any alternatives for this. We're using Riak 2.2.0.

**Comment by ovaltzer** [comment]
hey @arvind-kalyan ,
just run your way, worked after i recreated the same index with the modified schema

thanks

**Comment by asyncmind0** [comment]
Hi is there any update on this, all the methods in here seem very hacky. Is there a admin command or something in the works.

**Comment by rzezeski** [comment]
@jagguli Basho is dead. I believe there are some people (and companies) working on keeping Riak alive, but I doubt this feature will get done anytime soon (if ever). As much as I'd like to work on this I stay much too busy with other things to really dedicate the time.

@lastzero I look forward to using your database someday.

**Comment by zowers** [comment]
@rzezeski Basho is dead?
what does it mean? do you have proof?

**Comment by rzezeski** [comment]
@zowers Everyone was let go. The GitHub activity is all the proof you need.

**Comment by zowers** [comment]
ouch, that's sad to hear.
